test/test_model.py

```diff
@@ -3791,5 +3791,82 @@ def test_repr(self):
         )
 
 
+class TestCloudCredential(unittest.TestCase):
+    def setUp(self) -> None:
+        self.cloud_credential_dict = {
+            'auth-type': 'certificate',
+            'attrs': {
+                'client-cert': 'foo',
+                'client-key': 'bar',
+                'server-cert': 'baz'
+            },
+        }
+
+    def test_from_dict(self):
+        cloud_cred = ops.CloudCredential.from_dict(self.cloud_credential_dict)
+        self.assertEqual(cloud_cred.auth_type, 'certificate')
+        self.assertEqual(cloud_cred.attributes, self.cloud_credential_dict.get('attrs'))
+
+
+class TestCloudSpec(unittest.TestCase):
+    def setUp(self) -> None:
+        self.cloud_credential_dict = {
+            'auth-type': 'certificate',
+            'attrs': {
+                'client-cert': 'foo',
+                'client-key': 'bar',
+                'server-cert': 'baz'
+            },
+        }
+
+    def test_from_dict(self):
+        cloud_spec = ops.CloudSpec.from_dict(
+            {
+                'type': 'lxd',
+                'name': 'localhost',
+                'region': 'localhost',
+                'endpoint': 'https://10.76.251.1:8443',
+                'isControllerCloud': None,
+                'credential': self.cloud_credential_dict,
+                'identityEndpoint': None,
+                'storageEndpoint': None,
+                'caACertificates': None,
+                'skipTLSVerify': None
+            }
+        )
+        self.assertEqual(cloud_spec.type, 'lxd')
+        self.assertEqual(cloud_spec.name, 'localhost')
+        self.assertEqual(cloud_spec.region, 'localhost')
+        self.assertEqual(cloud_spec.endpoint, 'https://10.76.251.1:8443')
+        self.assertEqual(cloud_spec.is_controller_cloud, None)
+        self.assertEqual(cloud_spec.identity_endpoint, None)
+        self.assertEqual(cloud_spec.storage_endpoint, None)
+        self.assertEqual(cloud_spec.ca_certificates, None)
+        self.assertEqual(cloud_spec.skip_tls_verify, None)
+        self.assertEqual(
+            cloud_spec.credential,
+            ops.CloudCredential.from_dict(
+                self.cloud_credential_dict))
+
+
+class TestGetCloudSpec(unittest.TestCase):
+    def setUp(self):
+        self.model = ops.Model(ops.CharmMeta(), _ModelBackend('myapp/0'))
+
+    def test_get_cloud_spec(self):
+        fake_script(self, 'credential-get', """echo '{"type": "lxd", "name": "localhost"}'""")
+        cloud_spec = self.model.get_cloud_spec()
+        self.assertEqual(cloud_spec.type, 'lxd')
+        self.assertEqual(cloud_spec.name, 'localhost')
+        self.assertEqual(fake_script_calls(self, clear=True),
+                         [['credential-get', '--format=json']])
+
+    @patch("ops.model._ModelBackend.credential_get")
```

Comment:
I think we should do this one without mocking too, just like the above. Have the script `exit 1` or similar (there should be other similar examples in `test_model.py`).